Mathlib/AlgebraicGeometry/Noetherian.lean

```diff
@@ -0,0 +1,352 @@
+    IsLocallyNoetherian X ↔
+    ∀ (i : 𝒰.J), IsNoetherianRing Γ(𝒰.obj i, ⊤) := by
+  constructor
+  · intro h i
+    let U := Scheme.Hom.opensRange (𝒰.map i)
+    have := h.component_noetherian ⟨U, isAffineOpen_opensRange _⟩
+    apply isNoetherianRing_of_ringEquiv (R := Γ(X, U))
+    apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+    exact (Scheme.openImmersionΓ (𝒰.map i)).symm
+  · intro hCNoeth
+    let fS i : X.affineOpens := ⟨Scheme.Hom.opensRange (𝒰.map i), isAffineOpen_opensRange _⟩
+    let S : Set X.affineOpens := Set.range fS
+    apply isLocallyNoetherian_of_affine_cover (S := S)
+    rw [← Scheme.OpenCover.iSup_opensRange 𝒰, iSup_range]
+    rintro _ ⟨i, rfl⟩
+    apply isNoetherianRing_of_ringEquiv (R := Γ(𝒰.obj i, ⊤))
+    apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+    exact Scheme.openImmersionΓ (𝒰.map i)
+
+
+/-- A scheme is locally Noetherian if it has an open cover by
+locally Noetherian schemes. -/
+theorem isLocallyNoetherian_of_openCover (𝒰 : Scheme.OpenCover.{max v u, u} X)
+    (h𝒰 : ∀ (i : 𝒰.J), IsLocallyNoetherian (𝒰.obj i)) :
+    IsLocallyNoetherian X := by
+  let 𝒰' := Scheme.OpenCover.affineRefinement 𝒰
+  let m := Scheme.OpenCover.fromAffineRefinement 𝒰
+  apply Iff.mpr
+  apply isLocallyNoetherian_iff_of_affine_openCover (𝒰 := 𝒰'.openCover)
+  · intro i
+    rw [Scheme.AffineOpenCover.openCover_obj]
+    exact isAffine_Spec _
+  · intro i
+    let X := (𝒰.obj (m.idx i))
+    let U : X.affineOpens := ⟨(m.app i).opensRange, by
+      convert isAffineOpen_opensRange (m.app i)
+      exact isAffine_Spec _⟩
+    have hNoeth : IsNoetherianRing Γ(X, U) := by
+      apply (h𝒰 (m.idx i)).component_noetherian
+    apply isNoetherianRing_of_ringEquiv (R := Γ(X, U))
+    apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+    symm
+    exact Scheme.openImmersionΓ (m.app i)
+
+/-- If `R` is a noetherian ring, `Spec R` is a noetherian topological space. -/
+instance {R : CommRingCat} [IsNoetherianRing R] :
+    NoetherianSpace (Spec R) := by
+  convert PrimeSpectrum.instNoetherianSpace (R := R)
+
+lemma noetherianSpace_of_isAffine [IsAffine X]
+    (hX : IsNoetherianRing <| Γ(X, ⊤)) :
+    NoetherianSpace X := by
+  suffices f : Spec Γ(X, ⊤) ≃ₜ X by
+    apply (noetherianSpace_iff_of_homeomorph f).mp
+    infer_instance
+  apply TopCat.homeoOfIso
+  exact CategoryTheory.asIso (Scheme.isoSpec X).symm.hom.val.base
+
+lemma noetherianSpace_of_affineOpen (U : X.affineOpens)
+    (hU : IsNoetherianRing Γ(X, U)) :
+    NoetherianSpace U := by
+  suffices h : IsNoetherianRing Γ(X ∣_ᵤ ↑U, ⊤) by
+    exact noetherianSpace_of_isAffine h
+  apply isNoetherianRing_of_ringEquiv (R := Γ(X, U))
+  apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+  exact ((Scheme.restrictFunctorΓ X).app (op U)).symm
+
+/-- Any open immersion `Z ⟶ X` with `X` locally Noetherian is quasi-compact.
+
+[Stacks: Lemma 01OX](https://stacks.math.columbia.edu/tag/01OX) -/
+instance (priority := 100) {Z : Scheme} [IsLocallyNoetherian X]
```

Comment:
In some of the proofs below, you should use suitable indentation (and `·`) in order to highlight the hierarchy of the multiple goals.